tap_snowflake/client.py

```diff
@@ -74,15 +76,59 @@ class TableProfile:
 class SnowflakeConnector(SQLConnector):
     """Connects to the Snowflake SQL source."""
 
-    @classmethod
-    def get_sqlalchemy_url(cls, config: dict) -> str:
+    def get_private_key(self):
+        """Get private key from the right location."""
+
+        try:
+            encoded_passphrase = self.config["private_key_passphrase"].encode()
+        except KeyError:
+            encoded_passphrase = None
+
+        if "private_key_path" in self.config:
+            with open(self.config["private_key_path"], "rb") as key:
+                key_content = key.read()
+        else:
+            key_content = self.config["private_key"].encode()
+
+        p_key = serialization.load_pem_private_key(
+            key_content, password=encoded_passphrase, backend=default_backend()
+        )
+
+        return p_key.private_bytes(
+            encoding=serialization.Encoding.DER,
+            format=serialization.PrivateFormat.PKCS8,
+            encryption_algorithm=serialization.NoEncryption(),
+        )
+
+    @property
+    def auth_method(self):
+        """
+        Validate & return the authentication method based on config.
+
+        Cache computed auth_method to attribute `_auth_method`.
+        """
+        try:
+            return self._auth_method
+        except AttributeError:
```

Comment:
Would there be any downsides to instead do something like

```python
class SnowflakeConnector(SQLConnector):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._auth_method: str | None = None

    def auth_method(self) -> str:
        if self._auth_method:
            return self._auth_method
        ...
```

Reply:
No downsides as far as I can see, and I like yours as it avoids using exceptions for control flow, but it requires over-riding the init method. How do you feel about the following:

```
class SnowflakeConnector(SqlConnector):
  _auth_method: str | None = None

  def auth_method(self) -> str:
     if not self._auth_method:
       ...
     return self._auth_method
```

Reply:
> but it requires over-riding the init method

For my own curiosity, are there any downsides to doing that?

Reply:
no, as far as i can tell, the only downsides are additional lines one has to type, and every subsequent reader has to read - no downsides to performance or code organization.

actually, we can make the implementation even cleaner. New commit coming in a few minutes.

Reply:
@edgarrmondragon the new implementation uses `functools.cached_property`, thus we don't require declaring either a class or an instance variable to cache the value of this method evaluation. Also the if condition in the method can be removed making that code tighter as well.